src/util/transaction_identifier.h

```diff
@@ -0,0 +1,53 @@
+#ifndef BITCOIN_UTIL_TRANSACTION_IDENTIFIER_H
+#define BITCOIN_UTIL_TRANSACTION_IDENTIFIER_H
+
+#include <uint256.h>
+#include <util/types.h>
+
+#include <type_traits>
+
+/** transaction_identifier represents the two canonical transaction identifier
+ * types (txid, wtxid). */
+template <bool has_witness>
+class transaction_identifier : public uint256
+{
+    static constexpr bool negated_has_witness =
+        std::negation<std::bool_constant<has_witness>>::value;
+    // Opposite type w.r.t. `has_witness`
+    using opposite_type =
+        transaction_identifier<negated_has_witness>;
+
+public:
+    transaction_identifier() : uint256{} {}
+    explicit transaction_identifier(const uint256& other)
+        : uint256{other} {}
+
+    // Delete copy/move constructors/assignment for differing transaction_identifier types.
+    transaction_identifier(const opposite_type&) = delete;
+    transaction_identifier(opposite_type&&) = delete;
```

Comment:
> Alternatively it could be disallowed properly.

converting from `uint256` can now only be done using `{Txid,Wtxid}::FromUint256`. This required less changes then expected so I think that is actually preferable.

Reply:
Added a comment to `transaction_identifier` to note that it is not fully type-safe yet due to allowing implicit conversions to uint256 in the interim.